### ISSUE TITLE
Switch bwcli to using the new `bw serve` command

### ISSUE THREAD

**Issue by firecat53** [firecat53/bitwarden-menu#17]
Bitwarden CLI v1.22.0 added a `serve` option to interact with the vault using a [REST API](https://bitwarden.com/help/vault-management-api/). It seems faster than calling commands via subprocess, although need to wait to actually finish it publicly until they release the [fix](https://github.com/bitwarden/cli/pull/519) that binds the API to localhost. Right now it's wide open to the network.

Basic interaction:

    items = 'http://localhost:8087/unlock'
    headers = {'accept': 'application/json'}
    payload = {'password': 'mypassword'}
    res = requests.get(items, headers=headers, json=payload)

Probably start `bw serve` either in the main server process or perhaps in a separate thread that can be restarted with a different BITWARDENCLI_APPDATA_DIR to allow multiple vaults to be accessed.

**Comment by aclindsa** [comment]
I recently switched from keepass to bitwarden, and am searching for a replacement for keepmenu. My main issue with bitwarden-menu at the moment is that it takes a long time to initially open the database (it even seems to wait a second *before* prompting me for my password, though subsequent calls are quick enough).

Anyway, I'm curious if you think using the REST API as you describe here would reduce the startup time, or if you suspect the issue is elsewhere (it is possible I could be convinced to try implementing this if it isn't too difficult and is likely to reduce the startup time).

**Comment by aclindsa** [comment]
> I haven't made time to do any profiling as to where the significant delay is.

I just did some quick-n-dirty "profiling", and it seems like basically every call to the underlying CLI takes at least 2.5 seconds on my machine. When it takes 6 calls into the CLI on startup, you end up with at least 15 seconds of delay. Anyway, I've convinced myself that nearly all of the time is spent outside `bitwarden-menu` itself. I'm suspecting that most of the overhead is simply the nodejs startup cost, which does make me think its possible the `bw serve` command could help since we would only pay that startup cost 1x instead of 6x.

**Comment by firecat53** [comment]
Hmm, that implementation is disappointing. Thanks for testing, though! Well, it looks like they took your prompt and moved it to their engineering team for review, so that's promising! I'll keep an eye on that PR and see if they get it merged and then we can revisit. Really appreciate the work!!

**Comment by firecat53** [comment]
You're right. You have to login to the vault to start `bw serve` but any user can make requests to that port. Sad.

**Comment by Game4Move78** [comment]
Thanks for the help getting my PR noticed. In addition to CLI calls and server, you could use the internal API to download and decrypt the Bitwarden vault in Python using the `cryptography` module. I had a go at implementing this [here](https://github.com/Game4Move78/vaultio/blob/master/python/src/vaultio/vault/vault_sync.py), although some features are not implemented yet.

**Comment by aclindsa** [comment]
Thanks a ton @Game4Move78!

@firecat53 Given that it looks like this is going to go in upstream, I did a very quick and very hacky prototype to see if it seemed workable. Here's what I did so far:

https://github.com/firecat53/bitwarden-menu/compare/main...aclindsa:bitwarden-menu:main

I'm sure I'm breaking your code's assumptions left and right (for example, I didn't even attempt to handle switching vaults) and many API calls still fall back to the CLI, but this got me far enough to prove I could get the initial loading time down to only ~2 seconds!

I don't know how you prefer to handle stuff like this, but I'm more than happy for you to take what I have and bend it into something that's more in your coding style, for you to do it on your own, or whatever - just let me know.

**Comment by aclindsa** [comment]
I installed this on Arch Linux and it works great so far - Thanks, @firecat53! It cuts my initial load time down from around 12 seconds to 4 (measured after typing my password).

The one thing I noticed is that it seems to take about twice as much time as my earlier hack I've been using (~4 instead of 2 seconds between when I enter my password and when the menu is first displayed). I haven't spent time digging into this yet - I'm wondering if its simply because you are making an extra call into the server than my prototype did (perhaps necessarily?!). But I won't complain very loudly, because its already a great improvement!
